daffodil-core/src/main/scala/org/apache/daffodil/dpath/Conversions.scala

```diff
@@ -273,7 +273,7 @@ object Conversion {
           polymorphicExpressionDiagnostics(st, tt, context)
         context.SDE(
           "In expression %s, the type %s cannot be converted to %s.%s",
-          relevantExpr.text,
+          relevantExpr.wholeExpressionText,
```

Comment:
It looks like there are some other places where we reference ``.text`` an an Expression in a diagnostic in Expression.scala and Conversions.scala. Did you look at these and determine if a similar change to wholeExpressionText is needed to improve the diagnostics?

Reply:
Right, good call. I'll take a look and add them to the ticket.

Reply:
I addressed the additional case in Expressions.
The additional case in Conversions doesn't appear to be addressed by any unit test. The code appears to be largely duplicated in Expressions where the expression tests are applied.
If we need to develop a test to address the code in Conversion, I'll have some questions on how to construct the xsd.

Reply:
It would probably be worth adding additional tests, especially if we don't have any coverage. What questions do you have?

Reply:
I've been looking at it, and I don't see how to structure the test schema to trigger this section of the code. Didn't want to keep spinning my wheels.

Reply:
Gotcha. This bit of code should be hit if an expression ever needs to convert one type to another where that conversion isn't allowed. For example, it is illegal to convert an ``xs:date`` to an ``xs:boolean``. So you could have a schema that does something like this:

```xml
<xs:element name="foo" type="xs:boolean" dfdl:inputValueCalc="{ xs:boolean(xs:date("dateString")) }" />
```

That inputValueCalc expression includes an illegal conversion that *should* trigger this code.

Reply:
Perfect, I'll give that a try.

Reply:
So this issue of a new test for Conversions requires the generation of a StepExpression rather than a FunctionCallExpression. Can you give me an example of how to form a StepExpression in the schema.

Reply:
It's not clear to my why you need a step expression test. Is this because the original bug was about ``checkConstraints(.)``, which was about step expressions?

When you fixed the original bug by changing ``.text`` to ``.wholeExpression``, I suggested there were other places in code that could use this change that were not related to the original bug. So we sort of expanded the issue of this bug to fix the underlying issue, and not just this one specific issue of checkConstriants(.).

Conversions.scala doesn't really have anything specific to do with step expressions, but did have the incorrect .text. So I _think_ we just need a test that ensures this new conversion string outputs the full expression when this error occurs. There are probably existing tests that check illegal conversions, so if you could find those and just update the expected error message to ensure it includes the full expression. It might take some digging to find a test, so it might just be easier to create a new test.

Reply:
The error message in question is inside of case stepE: StepExpression => {} (Conversions LN 297). I placed a break point inside of this case and ran all tests. It is not hit. So, like you suggested it may make sense to create a new test to cover this code. Thus my effort to satisfy the StepExpression case in a new test. But it's possible I'm lost in a rabbit hole.

Reply:
I think I may be confused. Line 297 doesn't have an error message. Are you talking about the error message on line 317?

Reply:
Right, the case statement is on 297.

Reply:
Gotcha. This error is a bit trickier to hit. In order to hit this, there must be a global element that has an expression that references elements outside of it's global declaration. Depending on the context of where this global element is referenced, it could reference element with the same name but with different types. This is called a polymorphic expression, and isn't allowed. Here's a schema that triggers this error:

```xml
<?xml version="1.0" encoding="UTF-8"?> 

<xs:schema
  xmlns:xs="http://www.w3.org/2001/XMLSchema"
  xmlns:dfdl="http://www.ogf.org/dfdl/dfdl-1.0/">

  <xs:include schemaLocation="org/apache/daffodil/xsd/DFDLGeneralFormat.dfdl.xsd" />

  <xs:annotation>
    <xs:appinfo source="http://www.ogf.org/dfdl/">
      <dfdl:format ref="GeneralFormat"
        textBooleanTrueRep="t"
        textBooleanFalseRep="f" />
    </xs:appinfo>
  </xs:annotation>

  <xs:element name="root">
    <xs:complexType>
      <xs:sequence>
        <xs:element name="boolean">
          <xs:complexType>
            <xs:sequence>
              <xs:element name="bar" type="xs:boolean" dfdl:lengthKind="explicit" dfdl:length="1" />
              <xs:element ref="foo" />
            </xs:sequence>
          </xs:complexType>
        </xs:element>
        <xs:element name="date">
          <xs:complexType>
            <xs:sequence>
              <xs:element name="bar" type="xs:date" dfdl:lengthKind="explicit" dfdl:length="1" />
              <xs:element ref="foo" />
            </xs:sequence>
          </xs:complexType>
        </xs:element>
      </xs:sequence>
    </xs:complexType>
  </xs:element>

  <xs:element name="foo" type="xs:boolean" dfdl:inputValueCalc="{ xs:boolean(../bar) }" />

</xs:schema>
```
At the bottom we have a ``foo`` element that referenes a "../bar" element in the inputValueCalc expression, and converts ``../bar`` to a boolean. So what bar element is referenced must be able to be converted to a boolean.

In the root element, we then reference ``foo`` twice. In the first usage, the ``../bar`` that is referenced has an ``xs:boolean`` type and the conversion is allowed. But in the second usage, the ``../bar`` that is referenced has an ``xs:date`` type, which doesn't allow the conversion

Because ``../bar`` is referenced polymorphically (once to an xs:boolean, once to an xs:date) we should get the error you're looking for.

Reply:
A Step expression is one "step" in a Path expression. So foo/bar/baz/.././quux is a Path with 6 Step expressions in it.

Reply:
A step is either a named step (downward), or a ".." up step, or a "." self step.

Reply:
Thanks guys, very helpful.